### PR TITLE
SCSB-40 add JWST and RomanCal coverage tests

### PR DESCRIPTION
**Pull request by zacharyburnett**
https://jira.stsci.edu/browse/SCSB-40

organize tests into the following format:
image

### CODE REVIEW

**Comment by zacharyburnett** [comment]
several tests fail when testing via pipeline installations installed from source with `pip install git+`.

## failed tests in JWST pipeline
https://github.com/spacetelescope/stpipe/runs/6669190144?check_suite_focus=true
```
2022-05-31T11:20:51.4160233Z =================================== FAILURES ===================================
2022-05-31T11:20:51.4160700Z ___________________________ test_traverse_tso_grism ____________________________
2022-05-31T11:20:51.4161475Z [gw1] linux -- Python 3.9.12 /home/runner/work/stpipe/stpipe/.tox/jwst-cov/bin/python
2022-05-31T11:20:51.4161799Z 
2022-05-31T11:20:51.4161993Z     def test_traverse_tso_grism():
2022-05-31T11:20:51.4162690Z         """Make sure that the TSO dispersion polynomials are reversable."""
2022-05-31T11:20:51.4163096Z         wcsobj = create_tso_wcs()
2022-05-31T11:20:51.4163607Z         detector_to_grism = wcsobj.get_transform('direct_image', 'grism_detector')
2022-05-31T11:20:51.4164184Z         grism_to_detector = wcsobj.get_transform('grism_detector', 'direct_image')
2022-05-31T11:20:51.4164522Z     
2022-05-31T11:20:51.4164840Z         # TSGRISM always has same source locations
2022-05-31T11:20:51.4165281Z         # takes x,y,order -> ra, dec, wave, order
2022-05-31T11:20:51.4165623Z         xin, yin, order = (100, 100, 1)
2022-05-31T11:20:51.4165916Z     
2022-05-31T11:20:51.4166246Z         x0, y0, lam, orderdet = grism_to_detector(xin, yin, order)
2022-05-31T11:20:51.4166642Z         x, y, orderdet = detector_to_grism(x0, y0, lam, order)
2022-05-31T11:20:51.4166941Z     
2022-05-31T11:20:51.4167308Z         assert x0 == wcs_tso_kw['xref_sci']
2022-05-31T11:20:51.4167714Z >       assert y0 == wcs_tso_kw['yref_sci']
2022-05-31T11:20:51.4168050Z E       assert 56.393671399999874 == 35.0
2022-05-31T11:20:51.4168264Z 
2022-05-31T11:20:51.4168664Z .tox/jwst-cov/lib/python3.9/site-packages/jwst/assign_wcs/tests/test_nircam.py:213: AssertionError
2022-05-31T11:20:51.4169277Z ----------------------------- Captured stderr call -----------------------------
2022-05-31T11:20:51.4169869Z 2022-05-31 11:00:41,154 - stpipe.AssignWcsStep - INFO - AssignWcsStep instance created.
2022-05-31T11:20:51.4170447Z ------------------------------ Captured log call -------------------------------
2022-05-31T11:20:51.4170905Z INFO     stpipe.AssignWcsStep:step.py:330 AssignWcsStep instance created.
2022-05-31T11:20:51.4172588Z DEBUG    stpipe:nircam.py:45 reference files used in NIRCAM WCS pipeline: {'distortion': '~/crds_cache/references/jwst/nircam/jwst_nircam_distortion_0090.asdf', 'filteroffset': '~/crds_cache/references/jwst/nircam/jwst_nircam_filteroffset_0002.asdf', 'specwcs': '~/crds_cache/references/jwst/nircam/jwst_nircam_specwcs_0043.asdf', 'regions': None, 'wavelengthrange': '~/crds_cache/references/jwst/nircam/jwst_nircam_wavelengthrange_0002.asdf', 'camera': None, 'collimator': None, 'disperser': None, 'fore': None, 'fpa': None, 'msa': None, 'ote': None, 'ifupost': None, 'ifufore': None, 'ifuslicer': None}
2022-05-31T11:20:51.4174778Z DEBUG    stpipe:nircam.py:45 reference files used in NIRCAM WCS pipeline: {'distortion': '~/crds_cache/references/jwst/nircam/jwst_nircam_distortion_0090.asdf', 'filteroffset': '~/crds_cache/references/jwst/nircam/jwst_nircam_filteroffset_0002.asdf', 'specwcs': '~/crds_cache/references/jwst/nircam/jwst_nircam_specwcs_0043.asdf', 'regions': None, 'wavelengthrange': '~/crds_cache/references/jwst/nircam/jwst_nircam_wavelengthrange_0002.asdf', 'camera': None, 'collimator': None, 'disperser': None, 'fore': None, 'fpa': None, 'msa': None, 'ote': None, 'ifupost': None, 'ifufore': None, 'ifuslicer': None}
2022-05-31T11:20:51.4176072Z DEBUG    stpipe:nircam.py:143 Offsets from filteroffset file are 0.0, 0.0
2022-05-31T11:20:51.4176522Z DEBUG    stpipe:nircam.py:143 Offsets from filteroffset file are 0.0, 0.0
2022-05-31T11:20:51.4176940Z __________________________ test_traverse_wfss_grisms ___________________________
2022-05-31T11:20:51.4177491Z [gw0] linux -- Python 3.9.12 /home/runner/work/stpipe/stpipe/.tox/jwst-cov/bin/python
2022-05-31T11:20:51.4177973Z 
2022-05-31T11:20:51.4178164Z     def test_traverse_wfss_grisms():
2022-05-31T11:20:51.4178549Z         """Check the dispersion polynomials for each grism."""
2022-05-31T11:20:51.4179010Z         for pupil in ['GRISMR', 'GRISMC']:
2022-05-31T11:20:51.4179354Z >           traverse_wfss_trace(pupil)
2022-05-31T11:20:51.4179582Z 
2022-05-31T11:20:51.4179947Z .tox/jwst-cov/lib/python3.9/site-packages/jwst/assign_wcs/tests/test_nircam.py:196: 
2022-05-31T11:20:51.4180380Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-05-31T11:20:51.4180609Z 
2022-05-31T11:20:51.4180810Z pupil = 'GRISMR'
2022-05-31T11:20:51.4181014Z 
2022-05-31T11:20:51.4181281Z     def traverse_wfss_trace(pupil):
2022-05-31T11:20:51.4183540Z         """Make sure that the WFSS dispersion polynomials are reversable."""
2022-05-31T11:20:51.4183941Z         wcsobj = create_wfss_wcs(pupil)
2022-05-31T11:20:51.4184450Z         detector_to_grism = wcsobj.get_transform('detector', 'grism_detector')
2022-05-31T11:20:51.4184992Z         grism_to_detector = wcsobj.get_transform('grism_detector', 'detector')
2022-05-31T11:20:51.4185345Z     
2022-05-31T11:20:51.4185684Z         # check the round trip, grism pixel 100,100, source at 110,110,order 1
2022-05-31T11:20:51.4186097Z         xgrism, ygrism, xsource, ysource, orderin = (100, 100, 110, 110, 1)
2022-05-31T11:20:51.4186537Z         x0, y0, lam, order = grism_to_detector(xgrism, ygrism, xsource, ysource, orderin)
2022-05-31T11:20:51.4186982Z         x, y, xdet, ydet, orderdet = detector_to_grism(x0, y0, lam, order)
2022-05-31T11:20:51.4187317Z     
2022-05-31T11:20:51.4187579Z         assert x0 == xsource
2022-05-31T11:20:51.4187920Z >       assert y0 == ysource
2022-05-31T11:20:51.4188653Z E       assert 132.82171969999987 == 110
2022-05-31T11:20:51.4190031Z 
2022-05-31T11:20:51.4190404Z .tox/jwst-cov/lib/python3.9/site-packages/jwst/assign_wcs/tests/test_nircam.py:186: AssertionError
2022-05-31T11:20:51.4190934Z ----------------------------- Captured stderr call -----------------------------
2022-05-31T11:20:51.4191456Z 2022-05-31 11:00:41,674 - stpipe.AssignWcsStep - INFO - AssignWcsStep instance created.
2022-05-31T11:20:51.4191952Z ------------------------------ Captured log call -------------------------------
2022-05-31T11:20:51.4192349Z INFO     stpipe.AssignWcsStep:step.py:330 AssignWcsStep instance created.
2022-05-31T11:20:51.4193880Z DEBUG    stpipe:nircam.py:45 reference files used in NIRCAM WCS pipeline: {'distortion': '~/crds_cache/references/jwst/nircam/jwst_nircam_distortion_0090.asdf', 'filteroffset': '~/crds_cache/references/jwst/nircam/jwst_nircam_filteroffset_0002.asdf', 'specwcs': '~/crds_cache/references/jwst/nircam/jwst_nircam_specwcs_0059.asdf', 'regions': None, 'wavelengthrange': '~/crds_cache/references/jwst/nircam/jwst_nircam_wavelengthrange_0003.asdf', 'camera': None, 'collimator': None, 'disperser': None, 'fore': None, 'fpa': None, 'msa': None, 'ote': None, 'ifupost': None, 'ifufore': None, 'ifuslicer': None}
2022-05-31T11:20:51.4195944Z DEBUG    stpipe:nircam.py:45 reference files used in NIRCAM WCS pipeline: {'distortion': '~/crds_cache/references/jwst/nircam/jwst_nircam_distortion_0090.asdf', 'filteroffset': '~/crds_cache/references/jwst/nircam/jwst_nircam_filteroffset_0002.asdf', 'specwcs': '~/crds_cache/references/jwst/nircam/jwst_nircam_specwcs_0059.asdf', 'regions': None, 'wavelengthrange': '~/crds_cache/references/jwst/nircam/jwst_nircam_wavelengthrange_0003.asdf', 'camera': None, 'collimator': None, 'disperser': None, 'fore': None, 'fpa': None, 'msa': None, 'ote': None, 'ifupost': None, 'ifufore': None, 'ifuslicer': None}
2022-05-31T11:20:51.4196852Z DEBUG    stpipe:nircam.py:143 Offsets from filteroffset file are 0.0, 0.0
2022-05-31T11:20:51.4197221Z DEBUG    stpipe:nircam.py:143 Offsets from filteroffset file are 0.0, 0.0
```
## failed tests in Roman pipeline
https://github.com/spacetelescope/stpipe/runs/6669190260?check_suite_focus=true
```
2022-05-31T10:58:06.5877909Z =================================== FAILURES ===================================
2022-05-31T10:58:06.5878469Z _______________________________ test_entry_point _______________________________
2022-05-31T10:58:06.5879384Z [gw0] linux -- Python 3.9.12 /home/runner/work/stpipe/stpipe/.tox/romancal-cov/bin/python
2022-05-31T10:58:06.5879828Z 
2022-05-31T10:58:06.5881334Z     def test_entry_point():
2022-05-31T10:58:06.5931314Z         all_steps = entry_points.get_steps()
2022-05-31T10:58:06.5931641Z         roman_steps = [s for s in all_steps if s.package_name == "romancal"]
2022-05-31T10:58:06.5932021Z         tuples = {(s.class_name, s.class_alias, s.is_pipeline) for s in roman_steps}
2022-05-31T10:58:06.5932343Z >       assert tuples == set(get_steps())
2022-05-31T10:58:06.5932851Z E       AssertionError: assert set() == {('romancal.p..., False), ...}
2022-05-31T10:58:06.5933157Z E         Extra items in the right set:
2022-05-31T10:58:06.5933579Z E         ('romancal.pipeline.ExposurePipeline', 'roman_elp', True)
2022-05-31T10:58:06.5933993Z E         ('romancal.step.DQInitStep', None, False)
2022-05-31T10:58:06.5934379Z E         ('romancal.step.JumpStep', None, False)
2022-05-31T10:58:06.5934768Z E         ('romancal.step.RampFitStep', None, False)
2022-05-31T10:58:06.5935173Z E         ('romancal.step.DarkCurrentStep', None, False)
2022-05-31T10:58:06.5935578Z E         ('romancal.step.LinearityStep', None, False)
2022-05-31T10:58:06.5935974Z E         ('romancal.step.FlatFieldStep', None, False)
2022-05-31T10:58:06.5936354Z E         ('romancal.step.AssignWcsStep', None, False)
2022-05-31T10:58:06.5936751Z E         ('romancal.step.SaturationStep', None, False)
2022-05-31T10:58:06.5937142Z E         ('romancal.step.PhotomStep', None, False)
2022-05-31T10:58:06.5937404Z E         Full diff:
2022-05-31T10:58:06.5937609Z E         + set(,
2022-05-31T10:58:06.5937795Z E         + )
2022-05-31T10:58:06.5938026Z E         - {
2022-05-31T10:58:06.5938370Z E         -  ('romancal.pipeline.ExposurePipeline',
2022-05-31T10:58:06.5938704Z E         -   'roman_elp',
2022-05-31T10:58:06.5938960Z E         -   True),
2022-05-31T10:58:06.5939269Z E         -  ('romancal.step.AssignWcsStep',
2022-05-31T10:58:06.5939566Z E         -   None,
2022-05-31T10:58:06.5939812Z E         -   False),
2022-05-31T10:58:06.5940116Z E         -  ('romancal.step.DQInitStep',
2022-05-31T10:58:06.5940399Z E         -   None,
2022-05-31T10:58:06.5940629Z E         -   False),
2022-05-31T10:58:06.5940952Z E         -  ('romancal.step.DarkCurrentStep',
2022-05-31T10:58:06.5941247Z E         -   None,
2022-05-31T10:58:06.5941490Z E         -   False),
2022-05-31T10:58:06.5941805Z E         -  ('romancal.step.FlatFieldStep',
2022-05-31T10:58:06.5942080Z E         -   None,
2022-05-31T10:58:06.5942323Z E         -   False),
2022-05-31T10:58:06.5942622Z E         -  ('romancal.step.JumpStep',
2022-05-31T10:58:06.5942901Z E         -   None,
2022-05-31T10:58:06.5943141Z E         -   False),
2022-05-31T10:58:06.5943440Z E         -  ('romancal.step.LinearityStep',
2022-05-31T10:58:06.5943727Z E         -   None,
2022-05-31T10:58:06.5943969Z E         -   False),
2022-05-31T10:58:06.5944490Z E         -  ('romancal.step.PhotomStep',
2022-05-31T10:58:06.5944776Z E         -   None,
2022-05-31T10:58:06.5945019Z E         -   False),
2022-05-31T10:58:06.5945311Z E         -  ('romancal.step.RampFitStep',
2022-05-31T10:58:06.5945599Z E         -   None,
2022-05-31T10:58:06.5945840Z E         -   False),
2022-05-31T10:58:06.5946157Z E         -  ('romancal.step.SaturationStep',
2022-05-31T10:58:06.5946445Z E         -   None,
2022-05-31T10:58:06.5946673Z E         -   False),
2022-05-31T10:58:06.5946904Z E         - }
2022-05-31T10:58:06.5947030Z 
2022-05-31T10:58:06.5947391Z .tox/romancal-cov/lib/python3.9/site-packages/romancal/stpipe/tests/test_integration.py:28: AssertionError
```

**Comment by zacharyburnett** [comment]
it seems that the JWST tests actually fail in the repository:
https://github.com/spacetelescope/jwst/runs/6651735007?check_suite_focus=true
However, the Roman tests all pass in its repository:
https://github.com/spacetelescope/romancal/runs/6651691492?check_suite_focus=true

**Comment by hbushouse** [comment]
> it seems that the JWST tests actually fail in the repository: https://github.com/spacetelescope/jwst/runs/6651735007?check_suite_focus=true 

Yes, those JWST tests have been failing for nearly a week now due to some new NIRCam specwcs ref files delivered to CRDS last week. Still working on sorting out why we're getting those differences. Only happens when running tests against the CRDS OPS server for JWST, because that's the only place those new ref files are installed (for now).

**Comment by zacharyburnett** [comment]
After this PR is merged, the required status checks (`Settings` -> `Branches` -> `Branch Protection Rules` -> `master` -> `Required Status Checks`) should be updated to use the new names from these matrices:
- `test base library with coverage`
- `test (ubuntu-latest, py38)`
- `test (ubuntu-latest, py39)`
- `test (macos-latest, py39)`

File: tox.ini [hunk 1]
```diff
@@ -1,48 +1,94 @@
...
+    CI
+    CODECOV_*
+    CRDS_*
+    DISPLAY
+    HOME
+    TOXENV
+    TRAVIS
+    TRAVIS_*
+commands =
+    pip install -U stpipe==0.3.1
```
Comment (zacharyburnett) [inline tox.ini:91]: explicitly install `stpipe==0.3.1` (works for now, but does not test the current commit)

File: tox.ini [hunk 2]
```diff
@@ -1,48 +1,94 @@
...
-deps=
-    twine
-commands=
-    twine check {distdir}/*
+passenv =
+    HOME
+deps =
+    pytest-xdist
+commands =
+    pip freeze
```
Comment (zacharyburnett) [inline tox.ini:9]: this is just for console output

**Review by nden — APPROVED** [review]

**Comment by nden** [comment]
Could you redefine the  "required" tests?

**Pull request merged**